ngraph/core/reference/src/runtime/reference/convert.cpp

```diff
@@ -0,0 +1,273 @@
+                        movdqu(xword[rsp], f16vec);
+
+                        xor_(rsi, rsi);
+
+                        // Tail copying
+                        L(copy_loop);
+                        mov(di, word[rsp + rsi * sizeof(float16)]);
+                        mov(word[reg_dst], di);
+                        lea(reg_dst, ptr[reg_dst + sizeof(float16)]);
+                        inc(rsi);
+                        cmp(rsi, reg_sz);
+                        jl(copy_loop);
+
+                        add(rsp, 8 * sizeof(float)); // Free the array on stack
+
+                        L(exit);
+
+                        postamble();
+                    }
+
+                public:
+                    typedef struct
+                    {
+                        const uint8_t* arg;
+                        float16* out;
+                        const size_t count;
+                    } args_t;
+
+                    typedef void (*fn_t)(const args_t*);
+
+                    static fn_t get()
+                    {
+                        if (is_x64() && mayiuse(avx) && mayiuse(avx2) && mayiuse(fp16))
+                        {
+                            static jit_convert_u8_to_f16 generator;
+                            return (fn_t)generator.getCode();
+                        }
+                        return nullptr;
+                    }
+                };
+
+                class jit_convert_f16_to_f32 : public jit::Generator
+                {
+                    jit_convert_f16_to_f32()
+                    {
+                        using namespace Xbyak;
+
+                        auto f16vec = xmm1;
+                        auto f32vec = ymm2;
+
+                        auto reg_src = rax;
+                        auto reg_dst = rbx;
+                        auto reg_sz = rdx;
+
+                        Label loop, tail, tail_loop, copy_loop, exit;
+
+                        preamble();
+
+                        mov(reg_src, ptr[param + offsetof(args_t, arg)]);
+                        mov(reg_dst, ptr[param + offsetof(args_t, out)]);
+                        mov(reg_sz, ptr[param + offsetof(args_t, count)]);
+
+                        L(loop);
+
+                        cmp(reg_sz, 8);
+                        jl(tail);
+
+                        movdqu(f16vec, xword[reg_src]);
+                        vcvtph2ps(f32vec, f16vec);
+                        vmovups(yword[reg_dst], f32vec);
+
+                        lea(reg_src, ptr[reg_src + sizeof(float16) * 8]);
+                        lea(reg_dst, ptr[reg_dst + sizeof(float) * 8]);
+
+                        sub(reg_sz, 8);
+                        jmp(loop);
+
+                        L(tail);
+
+                        test(reg_sz, reg_sz);
+                        jz(exit);
+
+                        sub(rsp, 8 * sizeof(float)); // allocate array for 8 floats on stack
+                        xor_(rsi, rsi);
+
+                        mov(qword[rsp], rsi);
```

Comment:
Looks like you can just xor one vector register and fill the stack via single vector stor instruction.

Reply:
fixed